Add ssl support

m50d opened the pull request.

lmmworkv committed:
add ssl support

Thynix commented:
Cool, thanks! Could you please update `relaybot.config_sample` to demonstrate its usage?

Thynix commented on relaybot.py:
```diff
@@ -52,7 +53,10 @@ def get(option):
 
         factory = factory(options)
-        reactor.connectTCP(options['host'], int(options['port']), factory, int(options['timeout']))
+        if options['ssl']:
```
Thoughts on making `ssl` a boolean instead of the current specified-or-not?

m50d replied:
That'd probably be better. `boolean(options['ssl'])` would still be specified-or-not - what's the pythonic way to turn a string into a boolean in "do what I mean" fashion?

Thynix replied:
How about something like

```
mapping = {"True": True, "False": False}
def convert(value):
    if value in mapping:
        return mapping [value]
    raise TypeError("Cannot convert '{}' to boolean.".format(value))
```

I say that instead of just `value == "True"` to avoid something like "on"
becoming False.

m50d committed:
Tighten up ssl option parsing

Also add example to example config
Code untested as this machine doesn't have Python, but it looks correct.
The sentinel idiom is arguably overengineering.

wirehack7 commented:
It is not working:

```
File "relaybot.py", line 57, in main
            sentinel = Object()
        exceptions.NameError: global name 'Object' is not defined
```

Fixed: https://github.com/m50d/RelayBot/pull/1

Commit:
Fixed typo

Fixed typo of ```Object()``` to ```object()```

m50d committed:
Merge pull request #1 from wirehack7/patch-1

Fixed typo

Thynix commented on relaybot.py:
```diff
@@ -52,7 +53,15 @@ def get(option):
+        optionAsBoolean = { "": False, "false": False, "no": False, "true": True, "yes": True }
+        sentinel = object()
+        ssl = options.get('ssl', sentinel)
+        if(sentinel == ssl):
```
No need for parenthesis here, and `variable == value` seems preferable, so how about:

```
if ssl == sentinel:
```

If we're given that `options.get()` gives us the same default object, not just something equal to it, we could do:

```
if ssl is sentinel:
```

but I dunno if that's preferable.

Thynix replied:
Does it work better to go with explicit membership checks?

```
ssl = options['ssl']
if ssl in optionAsBoolean:
    if optionAsBoolean[ssl]:
        reactor.connectSSL(options['host'], int(options['port']), factory, ClientContextFactory(), int(options['timeout']))
    else:
        reactor.connectTCP(options['host'], int(options['port']), factory, int(options['timeout']))
else:
    raise TypeError("Cannot convert '{}' to boolean.".format(ssl))
```

m50d replied:
I'm used to `value == variable` checks just from C habit (makes accidental assignment less likely).
I prefer that to the explicit check but I don't especially mind one way or the other.

Thynix commented on relaybot.py:
```diff
@@ -52,7 +53,15 @@ def get(option):
-        reactor.connectTCP(options['host'], int(options['port']), factory, int(options['timeout']))
+        optionAsBoolean = { "": False, "false": False, "no": False, "true": True, "yes": True }
+        sentinel = object()
+        ssl = options.get('ssl', sentinel)
```
Should this be

```
optionAsBoolean.get(option['ssl'], sentinel)
```

?

m50d replied:
Yes, it should be

Thynix commented on relaybot.py:
```diff
@@ -52,7 +53,15 @@ def get(option):
 
         factory = factory(options)
-        reactor.connectTCP(options['host'], int(options['port']), factory, int(options['timeout']))
+        optionAsBoolean = { "": False, "false": False, "no": False, "true": True, "yes": True }
```
Empty string as `False` seems undesirable to me. We could also include "True" and "False" in case people use Python's boolean literal capitalization.

m50d replied:
I wanted it to come out as `False` if the option's not set, so as not to disrupt when people haven't configured it. Should that be `None: False`?

Thynix commented:
Sorry for the delay - I only got an email for the comment, and didn't get one when you posted changes.

m50d committed:
Update from review

m50d commented:
Have updated a bit - again I'm afraid I'm away from a machine that can actually run it. If you want to take over the feature rather than going back-and-forth with me that's fine, the "important" change was very simple, it's all the config-integration part that's taking effort.

Thynix commented:
Alright, will do. Thanks!

On Sat, Jun 6, 2015, 4:30 AM Michael Donaghy notifications@github.com
wrote:

> Have updated a bit - again I'm afraid I'm away from a machine that can
> actually run it. If you want to take over the feature rather than going
> back-and-forth with me that's fine, the "important" change was very simple,
> it's all the config-integration part that's taking effort.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/Thynix/RelayBot/pull/1#issuecomment-109552959.

Thynix commented:
At this point this fork is unmaintained. I'm sorry for letting your patch drop. [str4d's fork](https://github.com/str4d/RelayBot) looks like the most active one currently.
